controllers/provisioners/eksfargate/eksfargate.go

```diff
@@ -0,0 +1,290 @@
+		created, err := ctx.AwsWorker.CreateDefaultRole()
+		if created || err != nil {
+			ctx.Log.Info("Creating default role")
+			return err
+		}
+		role, err := ctx.AwsWorker.GetDefaultRole()
+		if err != nil {
+			ctx.Log.Info("Failed to get default role", "error", err)
+			return err
+		}
+		arn = *role.Arn
+
+		err = ctx.AwsWorker.AttachDefaultPolicyToDefaultRole()
+		if err != nil {
+			ctx.Log.Info("Failed to get attach policy to role", "error", err)
+			return err
+		}
+		ctx.Log.Info("Attached default policy to role")
+
+	} else {
+		arn = spec.GetPodExecutionRoleArn()
+	}
+
+	ctx.Log.Info("Creating a profile.", "arn", arn)
+	tryAgain, err := ctx.AwsWorker.CreateProfile(arn)
+	if tryAgain {
+		ctx.Log.Info("Resource inuse on Create.")
+		return nil
+	}
+	if err != nil {
+		ctx.Log.Info("Creation of the fargate profile failed", "cluster",
+			spec.GetClusterName(),
+			"profile",
+			spec.GetProfileName(),
+			"error", err)
+	} else {
+		instanceGroup.SetState(v1alpha1.ReconcileModifying)
+	}
+
+	return err
+}
+func (ctx *InstanceGroupContext) CloudDiscovery() error {
+	profile, err := ctx.AwsWorker.DescribeProfile()
+	if err != nil {
+		profile = &eks.FargateProfile{
+			Status: nil,
+		}
+	}
+	if profile.Status == nil {
+		ctx.DiscoveredState.ProfileStatus = aws.FargateProfileStatusMissing
+	} else {
+		ctx.DiscoveredState.ProfileStatus = *profile.Status
+	}
+	return nil
+}
+func (ctx *InstanceGroupContext) Delete() error {
+	instanceGroup := ctx.GetInstanceGroup()
+	spec := instanceGroup.GetEKSFargateSpec()
+
+	worker := ctx.AwsWorker
+	if spec.GetPodExecutionRoleArn() == "" {
+		found, err := worker.DetachDefaultPolicyFromDefaultRole()
+		if found || err != nil {
+			ctx.Log.Info("Detaching the default policy", "error", err)
+			return err
+		}
+
+		found, err = ctx.AwsWorker.DeleteDefaultRole()
+		if found || err != nil {
+			ctx.Log.Info("Deleting the default role", "error", err)
+			return err
+		}
+	}
+	ctx.Log.Info("Deleting the profile")
+	tryAgain, err := worker.DeleteProfile()
+	if tryAgain {
+		ctx.Log.Info("Resource inuse on Delete.")
+		return nil
+	}
+	if err != nil {
+		ctx.Log.Info("Deletion of the fargate profile.", "cluster",
+			spec.GetClusterName(),
+			"profile",
+			spec.GetProfileName(),
+			"error", err)
+		return err
+	}
+	instanceGroup.SetState(v1alpha1.ReconcileDeleting)
+
+	return err
+}
+
+func (ctx *InstanceGroupContext) Update() error {
+	// No update is required
+	ctx.Log.Info("Running update")
+	instanceGroup := ctx.GetInstanceGroup()
+	annos := instanceGroup.GetObjectMeta().GetAnnotations()
+	// If there is a last-applied-configuration then assume
+	// this is an update and throw an exception
+	if _, ok := annos[LastAppliedConfigurationKey]; ok {
+		return errors.New("update not supported")
+	}
+	instanceGroup.SetState(v1alpha1.ReconcileModified)
+	return nil
+}
+func (ctx *InstanceGroupContext) UpgradeNodes() error {
+	return errors.New("upgrade not supported")
```

Comment:
safer to return nil here instead of an error, even though this will never be called unless you set your state to InitUpgrade somewhere

Reply:
My reasoning is, if we get into UpgradeNodes, then there is definitely a bug and seems like an exception is the right response.  A nil response will hide it.

Reply:
But sure i can do it.  Just wanted you have you know my reasoning...

Reply:
Sure, I understand.
I think we prefer to 'err on the side of caution' and let the reconcile continue though, even if we accidentally add code that calls this method - otherwise the burden is on the user (he will get an error).